### PR TITLE
[Refactor] lock free with Locked key

### PR DESCRIPTION
**Pull request by rhdong**
- Change insert/assign/evict like API to use unique lock, and find use shared_lock
- Optimize the tests package name
- Fix: https://github.com/NVIDIA-Merlin/HierarchicalKV/issues/118

### CODE REVIEW

File: include/merlin/core_kernels.cuh [hunk 1]
```diff
@@ -749,6 +748,240 @@ __device__ __forceinline__ unsigned find_in_bucket(
...
+    }
+    result = g.shfl(result, src_lane);
+    if (result) {
+      // Not every `evicted_key` is correct expect the `src_lane` thread.
+      key_pos = g.shfl(local_min_meta_pos, src_lane);
+      return (evicted_key == static_cast<K>(RECLAIM_KEY))
+                 ? OccupyResult::OCCUPIED_RECLAIMED
+                 : OccupyResult::EVICT;
+    }
+  }
```
Comment (rhdong) [inline include/merlin/core_kernels.cuh:981]: Hi @Lifann , please help review here, thanks!

File: include/merlin/group_lock.hpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
+/*
```
Comment (rhdong) [inline include/merlin/group_lock.hpp:1]: Hi @zhangyafeikimi , could you help review on this group lock, thank you so much!

File: include/merlin/core_kernels.cuh [hunk 3]
```diff
@@ -749,6 +748,240 @@ __device__ __forceinline__ unsigned find_in_bucket(
...
+        src_lane = __ffs(vote) - 1;
+        key_pos = g.shfl(key_pos, src_lane);
+        return OccupyResult::DUPLICATE;
+      }
+      vote = g.ballot(expected_key == static_cast<K>(EMPTY_KEY));
+      if (vote) break;
+      vote = g.ballot(expected_key == static_cast<K>(LOCKED_KEY));
+    } while (vote != 0);
+
+    //     Step 2: (TBD)try find empty location.
```
Comment (Lifann) [inline include/merlin/core_kernels.cuh:827]: Too much blank space

File: include/merlin/group_lock.hpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  bool owns_lock() const noexcept { return owns_; }
+
+ private:
+  group_shared_mutex* const mutex_;
+  bool owns_;
+};
+
+class unique_lock {
```
Comment (zhangyafeikimi) [inline include/merlin/group_lock.hpp:177]: unique_lock -> write_read_lock
lock_both -> lock_write_read

File: include/merlin/group_lock.hpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+  ~unique_lock() {
+    if (owns_) {
+      mutex_->unlock_both();
+    }
+  }
+
+  void lock() {
+    if (owns_) {
+      throw(std::errc::resource_deadlock_would_occur);
```
Comment (zhangyafeikimi) [inline include/merlin/group_lock.hpp:198]: No throw, use an assertion instead.

File: include/merlin/group_lock.hpp [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+ private:
+  group_shared_mutex* const mutex_;
+  bool owns_;
+};
+
+class writer_shared_lock {
+ public:
+  writer_shared_lock(const writer_shared_lock&) = delete;
+  writer_shared_lock& operator=(const writer_shared_lock&) = delete;
```
Comment (zhangyafeikimi) [inline include/merlin/group_lock.hpp:147]: disable move ctor & move assignment

File: include/merlin/group_lock.hpp [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+ private:
+  std::atomic<int> writer_count_;
+  std::atomic<int> reader_count_;
+  std::atomic<bool> unique_flag_;
+};
+
+class reader_shared_lock {
+ public:
+  reader_shared_lock(const reader_shared_lock&) = delete;
+  reader_shared_lock& operator=(const reader_shared_lock&) = delete;
```
Comment (zhangyafeikimi) [inline include/merlin/group_lock.hpp:114]: disable move ctor & move assignment

File: include/merlin/group_lock.hpp [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+#include <system_error>
+#include <thread>
+
+namespace nv {
+namespace merlin {
+
+class group_shared_mutex {
+ public:
+  group_shared_mutex(const group_shared_mutex&) = delete;
+  group_shared_mutex& operator=(const group_shared_mutex&) = delete;
```
Comment (zhangyafeikimi) [inline include/merlin/group_lock.hpp:27]: disable move ctor & move assignment

File: include/merlin/group_lock.hpp [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+  explicit group_shared_mutex() noexcept
```
Comment (zhangyafeikimi) [inline include/merlin/group_lock.hpp:29]: explicit is redundant.

File: include/merlin/group_lock.hpp (hunk 2 above)
Comment (zhangyafeikimi) [reply include/merlin/group_lock.hpp:1]: have given some comments

File: include/merlin/group_lock.hpp (hunk 4 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp:177]: `unique_lock` follows the naming of `std::unique_lock` and would be easier to understand. Hope to keep the unique_lock, and will change `lock_both -> lock_write_read`.

File: include/merlin/group_lock.hpp (hunk 5 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp:198]: Per my knowledge, the assert is usually used for debug mode. We're trying to align the activity of group lock with the `std::shared_lock` and 'std::unique_lock', which throws the same exception. We can refer to https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/unique_lock.h#L138.

File: include/merlin/group_lock.hpp (hunk 6 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp:147]: Accept.

File: include/merlin/group_lock.hpp (hunk 7 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp:114]: Accept.

File: include/merlin/group_lock.hpp (hunk 8 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp:27]: Accept.

File: include/merlin/group_lock.hpp (hunk 9 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp:29]: Accept.

**Review by Lifann — APPROVED** [review]
LGTM

File: include/merlin/group_lock.hpp (hunk 4 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp]: Accept.

File: include/merlin/group_lock.hpp (hunk 5 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp]: Accept as we discuss offline.

File: include/merlin/group_lock.hpp (hunk 2 above)
Comment (rhdong) [reply include/merlin/group_lock.hpp:1]: Accept.

**Comment by rhdong** [comment]
Some issues of blossom caused a fail trigger CI workflow. We ran the CI testing manually and passed it successfully.

**Pull request merged**